ens/utils.py

```diff
@@ -23,38 +39,45 @@
 default = object()
 
 
-def Web3():
-    from web3 import Web3
-    return Web3
+if TYPE_CHECKING:
+    from web3 import Web3 as _Web3  # noqa: F401
+    from web3.providers import (  # noqa: F401
+        BaseProvider,
+    )
+
+
+def Web3() -> Type['_Web3']:
+    from web3 import Web3 as Web3Main
+    return Web3Main
 
 
-def dict_copy(func):
+def dict_copy(func: Callable[..., Any]) -> Callable[..., Any]:
```

Comment:
I think you need to use this pattern to preserve the return type of the function.

```
TFunc = TypeVar("TFunc", bound=Callable[..., Any])
def decorator(func: TFunc) -> TFunc:
    ...
```